web/src/main/kotlin/no/kartverket/matrikkel/bygning/plugins/Authentication.kt

```diff
@@ -1,60 +1,119 @@
 package no.kartverket.matrikkel.bygning.plugins
 
+
+import com.auth0.jwk.JwkProvider
 import com.auth0.jwk.JwkProviderBuilder
 import io.ktor.server.application.*
 import io.ktor.server.auth.*
 import io.ktor.server.auth.jwt.*
 import io.ktor.server.config.*
+import io.ktor.util.*
+import no.kartverket.matrikkel.bygning.config.Env
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 import java.net.URI
 import java.util.concurrent.TimeUnit
 
 private val log: Logger = LoggerFactory.getLogger(object {}::class.java)
 
-fun Application.configureMaskinportenAuthentication(config: ApplicationConfig) {
+fun Application.configureAuthentication(config: ApplicationConfig) {
     install(Authentication) {
         jwt("maskinporten") {
-            val shouldDisableMaskinporten = shouldDisableMaskinporten(config)
+            val isDisabled = isProviderDisabled(config, name!!)
 
-            skipWhen { shouldDisableMaskinporten }
+            skipWhen { isDisabled }
 
-            if (!shouldDisableMaskinporten) {
-                val authConfig = AuthenticationConfig(
-                    jwksUri = config.property("maskinporten.jwksUri").getString(),
-                    issuer = config.property("maskinporten.issuer").getString(),
-                    requiredScopes = config.property("maskinporten.scopes").getString(),
+            if (!isDisabled) {
+                val authConfig = JWTAuthenticationConfig(
+                    jwksUri = config.property("$name.jwksUri").getString(),
+                    issuer = config.property("$name.issuer").getString(),
+                    scopes = config.property("$name.scopes").getString(),
                 )
 
-                val jwkProvider = JwkProviderBuilder(URI(authConfig.jwksUri).toURL())
-                    .cached(10, 24, TimeUnit.HOURS)
-                    .rateLimited(10, 1, TimeUnit.MINUTES)
-                    .build()
-
-                verifier(jwkProvider, authConfig.issuer) {
+                verifier(authConfig.jwkProvider, authConfig.issuer) {
                     acceptLeeway(3)
-                    withClaim("scope", authConfig.requiredScopes)
+                    withClaim("scope", authConfig.scopes)
                 }
 
-                validate { it }
-
+                validate { it ->
+                    DigDirJWTPrincipal(it.payload.getClaim("orgno").asString())
+                }
             }
         }
+
+        jwtIdporten(config)
     }
+
 }
 
-private fun shouldDisableMaskinporten(config: ApplicationConfig): Boolean {
-    val shouldDisable = config.propertyOrNull("maskinporten.disabled")?.getString().toBoolean()
+data class DigDirJWTPrincipal(
```

Comment:
Usikker på om vi trenger å ha noe DigDir spesifikt i navnet her 🤔 Ser jo også for meg at dette potensielt kan være et EntraID token/principal også. Kanskje bare kalle det Principal..?

```suggestion
data class Principal(...)
```

Reply:
Endret det til Principal for nå, men skulle gjerne hatt _noe_ mer spesifikt enn det. Det er jo vi som oppretter en egen type for å være en umbrella principal for flere mer spesifikke typer JWTs per nå, som jo var DigDir per nå

Reply:
Ja, enig i det altså. Litt frem og tilbake, men kanskje DigDir er fint, det tydeliggjøre i hvert fall som du sier at principalet kommer fra "noe" fra DigDir og er mer custom enn bare et plain Principal.